test/distributed/fsdp/test_fsdp_summon_full_params.py

```diff
@@ -0,0 +1,176 @@
+
+        # we're assuming a single flatenned param
+        self.assertEqual(1, len(list(model.parameters())))
+
+        my_shard = torch.clone(next(model.parameters()))
+
+        with model._summon_full_params():
+            self.assertEqual(raw_model_size, self.get_model_param_count(model))
+            all_shards = next(model.parameters())
+            my_slice = torch.chunk(all_shards, self.world_size)[self.rank]
+
+            # shards are padded but the full_param tensor is not
+            self.assertTrue(torch.equal(my_shard[0: my_slice.numel()], my_slice))
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_recursive(self):
+        model = FSDP(nn.Sequential(
+            FSDP(nn.Linear(5, 5, bias=False)),
+            nn.Linear(5, 3, bias=False)
+        )).cuda(self.rank)
+
+        global_inner_numel = self.get_model_param_count(nn.Linear(5, 5, bias=False))
+        global_outer_numel = self.get_model_param_count(nn.Linear(5, 3, bias=False))
+
+        shard_inner_numel = int(math.ceil(global_inner_numel / self.world_size))
+        shard_outer_numel = int(math.ceil(global_outer_numel / self.world_size))
+
+        outer_param = model.get_parameter("_fsdp_wrapped_module.flat_param")
+        inner_param = model.get_parameter("_fsdp_wrapped_module._fpw_module.0._fsdp_wrapped_module.flat_param")
+        self.assertEqual(shard_outer_numel, outer_param.numel())
+        self.assertEqual(shard_inner_numel, inner_param.numel())
+
+        # non recursive, should only summon outer
+        with model._summon_full_params(recurse=False):
+            self.assertEqual(global_outer_numel, outer_param.numel())
+            self.assertEqual(shard_inner_numel, inner_param.numel())
+
+        # recursive, should summon everything
+        with model._summon_full_params(recurse=True):
+            self.assertEqual(global_outer_numel, outer_param.numel())
+            self.assertEqual(global_inner_numel, inner_param.numel())
+
+        # summon inner module only
+        with model[0]._summon_full_params(recurse=True):
+            self.assertEqual(shard_outer_numel, outer_param.numel())
+            self.assertEqual(global_inner_numel, inner_param.numel())
+
+        with model[0]._summon_full_params(recurse=False):
+            self.assertEqual(shard_outer_numel, outer_param.numel())
+            self.assertEqual(global_inner_numel, inner_param.numel())
+
+    @skip_if_lt_x_gpu(2)
+    def test_cannot_summon_full_params_from_forward(self):
+        class MyModule(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.a = nn.Parameter(torch.zeros(5))
+
+            def forward(self, fsdp_module):
+                with fsdp_module._summon_full_params():
+                    pass
+
+        model = FSDP(MyModule()).cuda(self.rank)
+        with self.assertRaisesRegex(ValueError, "current state is TrainingState_.FORWARD"):
+            model.forward(model)
```

Comment:
Usually to run the forward pass we want to just call `model(inputs)` as opposed to explicit call to `model.forward()`.

This is because `__call__` implementation does some other stuff (i.e. call hooks) prior to calling into forward and usually direct call to .forward() is discouraged in my experience.

Reply:
Good to know! Fixed.